### ISSUE TITLE
Simple Image Recognition using NotMNIST dataset

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#12111]
Simple Image Recognition using NotMNIST dataset<br>
<br>
https://ift.tt/zvkqYlo<br>
<br>
Lee Meng<br>
<br>

<div><div><div><div>
<div><pre><span></span><span>def</span> <span>make_arrays</span><span>(</span><span>nb_rows</span><span>,</span> <span>img_size</span><span>):</span>
  <span>if</span> <span>nb_rows</span><span>:</span>
    <span>dataset</span> <span>=</span> <span>np</span><span>.</span><span>ndarray</span><span>((</span><span>nb_rows</span><span>,</span> <span>img_size</span><span>,</span> <span>img_size</span><span>),</span> <span>dtype</span><span>=</span><span>np</span><span>.</span><span>float32</span><span>)</span>
    <span>labels</span> <span>=</span> <span>np</span><span>.</span><span>ndarray</span><span>(</span><span>nb_rows</span><span>,</span> <span>dtype</span><span>=</span><span>np</span><span>.</span><span>int32</span><span>)</span>
  <span>else</span><span>:</span>
    <span>dataset</span><span>,</span> <span>labels</span> <span>=</span> <span>None</span><span>,</span> <span>None</span>
  <span>return</span> <span>dataset</span><span>,</span> <span>labels</span>

<span>def</span> <span>merge_datasets</span><span>(</span><span>pickle_files</span><span>,</span> <span>train_size</span><span>,</span> <span>valid_size</span><span>=</span><span>0</span><span>):</span>
  <span>num_classes</span> <span>=</span> <span>len</span><span>(</span><span>pickle_files</span><span>)</span>
  <span>valid_dataset</span><span>,</span> <span>valid_labels</span> <span>=</span> <span>make_arrays</span><span>(</span><span>valid_size</span><span>,</span> <span>image_size</span><span>)</span>
  <span>train_dataset</span><span>,</span> <span>train_labels</span> <span>=</span> <span>make_arrays</span><span>(</span><span>train_size</span><span>,</span> <span>image_size</span><span>)</span>
  <span>vsize_per_class</span> <span>=</span> <span>valid_size</span> <span>//</span> <span>num_classes</span>
  <span>tsize_per_class</span> <span>=</span> <span>train_size</span> <span>//</span> <span>num_classes</span>
    
  <span>start_v</span><span>,</span> <span>start_t</span> <span>=</span> <span>0</span><span>,</span> <span>0</span>
  <span>end_v</span><span>,</span> <span>end_t</span> <span>=</span> <span>vsize_per_class</span><span>,</span> <span>tsize_per_class</span>
  <span>end_l</span> <span>=</span> <span>vsize_per_class</span><span>+</span><span>tsize_per_class</span>
  <span>for</span> <span>label</span><span>,</span> <span>pickle_file</span> <span>in</span> <span>enumerate</span><span>(</span><span>pickle_files</span><span>):</span>       
    <span>try</span><span>:</span>
      <span>with</span> <span>open</span><span>(</span><span>pickle_file</span><span>,</span> <span>'rb'</span><span>)</span> <span>as</span> <span>f</span><span>:</span>
        <span>letter_set</span> <span>=</span> <span>pickle</span><span>.</span><span>load</span><span>(</span><span>f</span><span>)</span>
        <span># let's shuffle the letters to have random validation and training set</span>
        <span>np</span><span>.</span><span>random</span><span>.</span><span>shuffle</span><span>(</span><span>letter_set</span><span>)</span>
        <span>if</span> <span>valid_dataset</span> <span>is</span> <span>not</span> <span>None</span><span>:</span>
          <span>valid_letter</span> <span>=</span> <span>letter_set</span><span>[:</span><span>vsize_per_class</span><span>,</span> <span>:,</span> <span>:]</span>
          <span>valid_dataset</span><span>[</span><span>start_v</span><span>:</span><span>end_v</span><span>,</span> <span>:,</span> <span>:]</span> <span>=</span> <span>valid_letter</span>
          <span>valid_labels</span><span>[</span><span>start_v</span><span>:</span><span>end_v</span><span>]</span> <span>=</span> <span>label</span>
          <span>start_v</span> <span>+=</span> <span>vsize_per_class</span>
          <span>end_v</span> <span>+=</span> <span>vsize_per_class</span>
                    
        <span>train_letter</span> <span>=</span> <span>letter_set</span><span>[</span><span>vsize_per_class</span><span>:</span><span>end_l</span><span>,</span> <span>:,</span> <span>:]</span>
        <span>train_dataset</span><span>[</span><span>start_t</span><span>:</span><span>end_t</span><span>,</span> <span>:,</span> <span>:]</span> <span>=</span> <span>train_letter</span>
        <span>train_labels</span><span>[</span><span>start_t</span><span>:</span><span>end_t</span><span>]</span> <span>=</span> <span>label</span>
        <span>start_t</span> <span>+=</span> <span>tsize_per_class</span>
        <span>end_t</span> <span>+=</span> <span>tsize_per_class</span>
    <span>except</span> <span>Exception</span> <span>as</span> <span>e</span><span>:</span>
      <span>print</span><span>(</span><span>'Unable to process data from'</span><span>,</span> <span>pickle_file</span><span>,</span> <span>':'</span><span>,</span> <span>e</span><span>)</span>
      <span>raise</span>
    
  <span>return</span> <span>valid_dataset</span><span>,</span> <span>valid_labels</span><span>,</span> <span>train_dataset</span><span>,</span> <span>train_labels</span>
            
            
<span>train_size</span> <span>=</span> <span>200000</span>
<span>valid_size</span> <span>=</span> <span>10000</span>
<span>test_size</span> <span>=</span> <span>10000</span>

<span>valid_dataset</span><span>,</span> <span>valid_labels</span><span>,</span> <span>train_dataset</span><span>,</span> <span>train_labels</span> <span>=</span> <span>merge_datasets</span><span>(</span>
  <span>train_datasets</span><span>,</span> <span>train_size</span><span>,</span> <span>valid_size</span><span>)</span>
<span>_</span><span>,</span> <span>_</span><span>,</span> <span>test_dataset</span><span>,</span> <span>test_labels</span> <span>=</span> <span>merge_datasets</span><span>(</span><span>test_datasets</span><span>,</span> <span>test_size</span><span>)</span>

<span>print</span><span>(</span><span>'Training:'</span><span>,</span> <span>train_dataset</span><span>.</span><span>shape</span><span>,</span> <span>train_labels</span><span>.</span><span>shape</span><span>)</span>
<span>print</span><span>(</span><span>'Validation:'</span><span>,</span> <span>valid_dataset</span><span>.</span><span>shape</span><span>,</span> <span>valid_labels</span><span>.</span><span>shape</span><span>)</span>
<span>print</span><span>(</span><span>'Testing:'</span><span>,</span> <span>test_dataset</span><span>.</span><span>shape</span><span>,</span> <span>test_labels</span><span>.</span><span>shape</span><span>)</span>
</pre></div>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via LeeMeng - All About DataScience <br>
<br>
April 21, 2025 at 09:35AM
